### ISSUE TITLE
store encryption key with Android's keystore

### ISSUE THREAD

**Issue by nioncode** [spesmilo/electrum#6846]
If I'm not mistaken, electrum on Android does NOT use Android's `KeyStore` to store the wallet's encryption key, but derives an encryption key from the wallet passphrase that is used to decrypt the wallet. Is my understanding correct?

If this is the case, then I strongly suggest that the encryption key should be stored with Android's `KeyStore`, which is backed by hardware security for many Android devices and is thus significantly harder to extract or bruteforce than the key derivation used by electrum (which can be performed via an offline attack on much more powerful hardware after getting hold of the encrypted wallet).

**Comment by SomberNight** [comment]
Note that the wallet file is stored in private storage, so the attacker would need root to get access to it.

**Comment by nioncode** [comment]
That is true, yes, but there are quite a few exploits available depending on the used Android version and the patchlevel, so getting hold of the wallet file is theoretically possible and then it could be bruteforced, since most users don't use strong passwords on mobile devices.

If the encryption key would be stored in the Android `KeyStore`, then the key would be protected by the lockscreen password, which is subject to rate limitations and cannot easily be bruteforced.

That being said, the attack scenario is still quite complex and not an immediate threat. If the user's phone is exploited at runtime, then you have lost already anyways and the attacker can also modify the installed electrum app to leak the private key (or simply sweep the funds).
